File: src/discovery/docker.ts
export interface DockerContainer {
  id: string;
  name: string;
  image: string;
  ports: number[];
  ip: string | null;
  network: string | null;
  workdir: string | null;
  labels: Record<string, string>;
}

interface DockerPsOutput {
  ID: string;
  Names: string;
  Image: string;
  Ports: string;
  Labels: string;
}

interface DockerInspectOutput {
  NetworkSettings: {
    Networks: Record<
      string,
      {
        IPAddress: string;
      }
    >;
  };
  Config: {
    Labels: Record<string, string>;
    ExposedPorts?: Record<string, unknown>;
    WorkingDir?: string;
  };
}

// Get the compose project from environment variable
function getOwnComposeProject(): string | null {
  return process.env.COMPOSE_PROJECT || null;
}

export async function discoverDockerContainers(): Promise<DockerContainer[]> {
  const containers: DockerContainer[] = [];

  // Check if docker is available
  try {
    const checkProc = Bun.spawn(["docker", "info"], {
      stdout: "pipe",
      stderr: "pipe",
    });
    await checkProc.exited;
    if (checkProc.exitCode !== 0) {
      return [];
    }
  } catch {
    return [];
  }

  // Get our own compose project to filter it out
  const ownProject = getOwnComposeProject();

  // Get running containers
  const psProc = Bun.spawn(
    ["docker", "ps", "--format", "{{json .}}"],
    {
      stdout: "pipe",
      stderr: "pipe",
    }
  );

  const psOutput = await new Response(psProc.stdout).text();
  const psLines = psOutput.split("\n").filter(Boolean);

  for (const line of psLines) {
    try {
      const container: DockerPsOutput = JSON.parse(line);
      const details = await getContainerDetails(container.ID);

      if (details) {
        // Filter out containers from our own compose project
        const containerProject = details.labels["com.docker.compose.project"];
        if (ownProject && containerProject === ownProject) {
          continue;
        }

        containers.push(details);
      }
    } catch {
      // Skip malformed JSON
    }
  }

  return containers;
}

async function getContainerDetails(
  containerId: string
): Promise<DockerContainer | null> {
  const inspectProc = Bun.spawn(["docker", "inspect", containerId], {
    stdout: "pipe",
    stderr: "pipe",
  });

  const output = await new Response(inspectProc.stdout).text();

  try {
    const inspectData: DockerInspectOutput[] = JSON.parse(output);
    const data = inspectData[0];
    if (!data) return null;

    const networks = data.NetworkSettings.Networks;
    const labels = data.Config.Labels || {};

    // Get first available network and IP
    let ip: string | null = null;
    let network: string | null = null;
    for (const [netName, netConfig] of Object.entries(networks)) {
      if (netConfig.IPAddress) {
        ip = netConfig.IPAddress;
        network = netName;
        break;
      }
    }

    // Extract exposed ports
    const ports: number[] = [];
    if (data.Config.ExposedPorts) {
      for (const portSpec of Object.keys(data.Config.ExposedPorts)) {
        const portMatch = portSpec.match(/^(\d+)/);
        if (portMatch?.[1]) {
          ports.push(parseInt(portMatch[1], 10));
        }
      }
    }

    // Get container name from inspect
    const nameProc = Bun.spawn(
      ["docker", "inspect", "--format", "{{.Name}}", containerId],
      {
        stdout: "pipe",
        stderr: "pipe",
      }
    );
    const nameOutput = await new Response(nameProc.stdout).text();
    const name = nameOutput.trim().replace(/^\//, "");

    // Get image name
    const imageProc = Bun.spawn(
      ["docker", "inspect", "--format", "{{.Config.Image}}", containerId],
      {
        stdout: "pipe",
        stderr: "pipe",
      }
    );
    const imageOutput = await new Response(imageProc.stdout).text();
    const image = imageOutput.trim();

    // Get workdir - prefer docker-compose working_dir label, then container's WorkingDir
    const workdir =
      labels["com.docker.compose.project.working_dir"] ||
      data.Config.WorkingDir ||
      null;

    return {
      id: containerId,
      name,
      image,
      ports,
      ip,
      network,
      workdir,
      labels,
    };
  } catch {
    return null;
  }
}

export async function getContainerIp(
  containerIdOrName: string
): Promise<string | null> {
  const proc = Bun.spawn(
    [
      "docker",
      "inspect",
      "--format",
      "{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}",
      containerIdOrName,
    ],
    {
      stdout: "pipe",
      stderr: "pipe",
    }
  );

  const output = await new Response(proc.stdout).text();
  const ip = output.trim();
  return ip || null;
}

File: llm_resolver/discovery/docker.go
package discovery

import (
	"context"
	"encoding/json"
	"fmt"
	"os/exec"
	"regexp"
	"strings"
	"time"
)

const commandTimeout = 10 * time.Second

// Pre-compiled regex for port extraction
var portRegex = regexp.MustCompile(`^(\d+)`)

// DockerContainer represents a discovered Docker container
type DockerContainer struct {
	ID      string            `json:"id"`
	Name    string            `json:"name"`
	Image   string            `json:"image"`
	Ports   []int             `json:"ports"`
	IP      string            `json:"ip"`
	Network string            `json:"network"`
	Workdir string            `json:"workdir"`
	Labels  map[string]string `json:"labels"`
}

// dockerPsOutput represents the JSON output from docker ps
type dockerPsOutput struct {
	ID     string `json:"ID"`
	Names  string `json:"Names"`
	Image  string `json:"Image"`
	Ports  string `json:"Ports"`
	Labels string `json:"Labels"`
}

// dockerInspectOutput represents the relevant parts of docker inspect output
type dockerInspectOutput struct {
	Name            string `json:"Name"`
	NetworkSettings struct {
		Networks map[string]struct {
			IPAddress string `json:"IPAddress"`
		} `json:"Networks"`
	} `json:"NetworkSettings"`
	Config struct {
		Image        string            `json:"Image"`
		Labels       map[string]string `json:"Labels"`
		ExposedPorts map[string]struct{}
		WorkingDir   string `json:"WorkingDir"`
	} `json:"Config"`
}

// DiscoverDockerContainers discovers running Docker containers
func DiscoverDockerContainers(ownComposeProject string) ([]DockerContainer, error) {
	ctx, cancel := context.WithTimeout(context.Background(), commandTimeout)
	defer cancel()

	// Check if docker is available
	if err := exec.CommandContext(ctx, "docker", "info").Run(); err != nil {
		return nil, nil // Docker not available, return empty list
	}

	// Get running containers
	cmd := exec.CommandContext(ctx, "docker", "ps", "--format", "{{json .}}")
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	var containers []DockerContainer

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}

		var ps dockerPsOutput
		if err := json.Unmarshal([]byte(line), &ps); err != nil {
			continue
		}

		// Get detailed container info
		details, err := getContainerDetails(ps.ID)
		if err != nil || details == nil {
			continue
		}

		// Filter out containers from our own compose project
		if ownComposeProject != "" {
			containerProject := details.Labels["com.docker.compose.project"]
			if containerProject == ownComposeProject {
				continue
			}
		}

		containers = append(containers, *details)
	}

	return containers, nil
}

// getContainerDetails gets detailed information about a container
func getContainerDetails(containerID string) (*DockerContainer, error) {
	ctx, cancel := context.WithTimeout(context.Background(), commandTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "docker", "inspect", containerID)
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	var inspectData []dockerInspectOutput
	if err := json.Unmarshal(output, &inspectData); err != nil {
		return nil, err
	}

	if len(inspectData) == 0 {
		return nil, nil
	}

	data := inspectData[0]

	// Get first available network and IP
	var ip, network string
	for netName, netConfig := range data.NetworkSettings.Networks {
		if netConfig.IPAddress != "" {
			ip = netConfig.IPAddress
			network = netName
			break
		}
	}

	// Extract exposed ports
	var ports []int
	for portSpec := range data.Config.ExposedPorts {
		if match := portRegex.FindStringSubmatch(portSpec); len(match) > 1 {
			if port, err := parsePort(match[1]); err == nil {
				ports = append(ports, port)
			}
		}
	}

	// Get workdir - prefer docker-compose working_dir label, then container's WorkingDir
	workdir := data.Config.Labels["com.docker.compose.project.working_dir"]
	if workdir == "" {
		workdir = data.Config.WorkingDir
	}

	// Clean container name (remove leading /)
	name := strings.TrimPrefix(data.Name, "/")

	return &DockerContainer{
		ID:      containerID,
		Name:    name,
		Image:   data.Config.Image,
		Ports:   ports,
		IP:      ip,
		Network: network,
		Workdir: workdir,
		Labels:  data.Config.Labels,
	}, nil
}

// GetContainerIP gets the IP address of a container by name or ID
func GetContainerIP(containerIDOrName string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), commandTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "docker", "inspect", "--format", "{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}", containerIDOrName)
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(output)), nil
}

// parsePort parses and validates a port string, returning error for invalid input
func parsePort(s string) (int, error) {
	if s == "" {
		return 0, fmt.Errorf("empty port string")
	}
	var port int
	for _, c := range s {
		if c < '0' || c > '9' {
			return 0, fmt.Errorf("invalid character in port: %c", c)
		}
		port = port*10 + int(c-'0')
		if port > 65535 {
			return 0, fmt.Errorf("port out of range: %d", port)
		}
	}
	if port < 1 {
		return 0, fmt.Errorf("port must be >= 1")
	}
	return port, nil
}
